ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ChainSync/Client/HistoricalRollbacks.hs

```diff
@@ -0,0 +1,134 @@
+      -- | The maximum age of a rollback at arrival time.
+    , maxRollbackAge :: !NominalDiffTime
+    }
+  deriving anyclass (Exception)
+
+deriving stock instance Show HistoricalRollbackException
+
+instance Eq HistoricalRollbackException where
+  (==)
+    (HistoricalRollbackException (l0 :: Point l) l1 l2 l3)
+    (HistoricalRollbackException (r0 :: Point r) r1 r2 r3)
+    = case eqT @l @r of
+        Nothing   -> False
+        Just Refl -> (l0, l1, l2, l3) == (r0, r1, r2, r3)
+
+{-------------------------------------------------------------------------------
+  Real implmementation
+-------------------------------------------------------------------------------}
+
+-- | Do not perform any historicity checks. This is useful when we only sync
+-- from trusted peers (Praos mode) or when the impact of historical rollbacks is
+-- already mitigated by other means (for example the Limit on Patience in the
+-- case of Genesis /without/ ChainSync Jumping).
+noCheck :: Applicative m => HistoricalRollbackCheck m blk
+noCheck = HistoricalRollbackCheck {
+      onRollBackward = \_lcfg _lst _rollbackPoint -> pure $ Right ()
+    }
+
+-- | Deny all rollbacks that are older than the given 'NominalDiffTime' upon
+-- arrival.
+mkCheck ::
+     forall m blk.
+     ( Monad m
+     , HasHeader blk
+     , HasHardForkHistory blk
+     )
+  => SystemTime m
+  -> m GsmState
+     -- ^ Get the current 'GsmState'.
+     --
+     -- This is used to disable the historical rollback check when we are caught
+     -- up. The rationale is extra resilience against disconnects between honest
+     -- nodes in disaster scenarios with very low chain density.
+  -> NominalDiffTime
+     -- ^ The maximum age of a rollback at arrival time.
+     --
+     -- This should be set to at least the maximum duration duration (across all
+     -- eras) of a stability window.
+  -> HistoricalRollbackCheck m blk
+mkCheck systemTime getCurrentGsmState maxRollbackAge = HistoricalRollbackCheck {
+      onRollBackward = \lcfg lst rollbackPoint -> getCurrentGsmState >>= \case
+          PreSyncing -> judgeRollback lcfg lst rollbackPoint
+          Syncing    -> judgeRollback lcfg lst rollbackPoint
+          CaughtUp   -> pure $ Right ()
+    }
+  where
+    judgeRollback lcfg lst rollbackPoint = do
+        arrivalTime <- systemTimeCurrent systemTime
+        pure $ when (maxRollbackAge < arrivalTime `diffRelTime` slotTime) $
+          throwError HistoricalRollbackException {
+              rollbackPoint
+            , slotTime
+            , arrivalTime
+            , maxRollbackAge
+            }
+      where
+        -- Use slot 0 in case we are rolling back to Genesis.
+        rollbackSlot  = fromWithOrigin 0 $ pointSlot rollbackPoint
+        qry           = Qry.slotToWallclock rollbackSlot
+        hfSummary     = hardForkSummary lcfg lst
+        -- This can't fail as per the precondition on 'onRollBackward'.
+        (slotTime, _) = Qry.runQueryPure qry hfSummary
```

Comment:
I think there's a different approach that would work and maybe be simpler/clearer, but would require some yak shaving.

The idea would be to record the UTC time of a header's slot when validating that header. We necessarily have it, because of the future check. So maybe we could simply add it as a field of the `HeaderState` data type, for example.

Then the roll back handler (ie `HeaderState.rollback`) could return the oldest header state that was rolled back (if any). This check here would merely need to compare that oldest-rolled-back UTC time against the current wall clock.

Moreover: this would actually avoid the required-msg-rollback-after-intersection wrinkle since such a `MsgRollBack` doesn't actually rollback any blocks. (And if it somehow did, that'd be a proper roll back, not dismissable as just part of an intersection response.)

Reply:
Hmm, this sounds interesting; it would of course invalidate existing ledger snapshots (or require some rather ad-hoc code for migration).

> Then the roll back handler (ie `HeaderState.rollback`) could return the oldest header state that was rolled back (if any). This check here would merely need to compare that oldest-rolled-back UTC time against the current wall clock.

Any particular reason for not using the `HeaderState` of the tip of the result of `HeaderState.rollback`? Just because it would avoid the "required-msg-rollback-after-intersection wrinkle"?

> Moreover: this would actually avoid the required-msg-rollback-after-intersection wrinkle since such a `MsgRollBack` doesn't actually rollback any blocks. (And if it somehow did, that'd be a proper roll back, not dismissable as just part of an intersection response.)

This would be equivalent to just not doing the historical rollback check if the rollback isn't actually rolling back any blocks even in the current state of the PR, right?

Reply:
The PR has been rewritten to this approach: the payload of the `HeaderStateHistory` is now `HeaderStateWithTime` which is a `HeaderState` and a `RelativeTime` of the header slot. This approach avoids invalidating ledger snapshots.